### ISSUE TITLE
DllNotFoundException

### ISSUE THREAD

**Issue by m21c** [pollen-robotics/reachy2021-unity-package#11]
Salut! 
I'm just getting started with Reachy on Unity (as I don't have the physical robot), and after following the instructions on the Readme (and on [your Medium article](https://medium.com/pollen-robotics/controlling-a-reachy-robot-in-unity-f3d90d550345)), when I click on Play, I am getting the below error (also see the attached screenshot):
```
DllNotFoundException: libArm_kinematics.so assembly:<unknown assembly> type:<unknown type> member:(null)
JointServiceServer.Start () (at Assets/ReachySimulator/Scripts/JointServiceServer.cs:56)
```
I previously tested the 2019 version (always on Unity) and thanks to the WsIO script I managed to connect to Reachy via Python commands. Should I add a similar script in Reachy's inspector panel? If not, could you please advise me on how to fix this? 

Another question: is it possible to use Reachy's cameras on Unity? What would be the host address? I'd like to add a webcam feed to that.

Merci bien!
Miriam

![image](https://user-images.githubusercontent.com/101079942/179669448-6d02a051-6fa5-4bfa-8129-38df734d4094.png)

**Comment by FabienDanieau** [comment]
Bonjour Miriam :)

Can you confirm that libArm_kinematics.so is present in the folder Packages/ReachySimulator/Plugins/ ?
Do you also have the GRPC dll in Assets/Plugins? (folders Google.Protobuf, Grpc.Core, System.Memory, etc) ?

Regarding your question about the cameras, you can get the frames of the virtual cameras is that what you want to do?

**Comment by FabienDanieau** [comment]
It is ok to have this structure. It is because you have imported the package directly. ReachySimulator would be in the Packages folder if you use the package manager to install it. Both are equivalent.
The libArm_kinematics.so should be found, but maybe it refers to a library that is not on your system. Could you use dependency walker (https://dependencywalker.com/) to list the missing libraries (if any)? Just open this software and, drag n drop the lib. Thanks

**Comment by m21c** [comment]
I'm actually working on Ubuntu, and dependency worker seems to be only working on Windows... How can I do it? 

I think I managed to have the connection working, so I can now send commands, and the virtual Reachy works. 

However, I'm still getting the libArm_kinematics.so error. In fact, I'm struggling with using the forward and inverse kinematics functions. When I follow [these instructions](https://docs.pollen-robotics.com/sdk/first-moves/kinematics/#forward-kinematics), I'm getting this error: 

![image](https://user-images.githubusercontent.com/101079942/179930467-4d2df81e-9283-44b5-8c9c-dd34bc2a7600.png)

Thanks

**Comment by m21c** [comment]
Thank you very much, Fabien! That was indeed the issue. So, just for the records (in case anyone else comes across this same problem), my missing library was `GLIBCXX_3.4.29`: 

```
$ ldd libArm_kinematics.so

./libArm_kinematics.so: /lib/x86_64-linux-gnu/libstdc++.so.6: version 'GLIBCXX_3.4.29' not found (required by ./libArm_kinematics.so)
    linux-vdso.so.1 (0x00007ffced5c4000)
    libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f435394f000)
    libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4353800000)
    libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f43537e5000)
    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f43535f3000)
    /lib64/ld-linux-x86-64.so.2 (0x00007f4353c37000)
```


I installed the libstdc++6 as suggested [here](https://askubuntu.com/questions/575505/glibcxx-3-4-20-not-found-how-to-fix-this-error). Then, I closed Unity project, re-opened it again, and the error was no longer there. In fact, even from python, I can now run the forward and inverse kinematics nice and easy! 

Merci beacoup!

**Comment by FabienDanieau** [comment]
Nice! Good to know. I'll close the issue. Feel free to open another one for the camera if you still have a problem.
BTW, if you try the example provided in the [Medium article](https://medium.com/pollen-robotics/controlling-a-reachy-robot-in-unity-f3d90d550345) with the basket ball, you'll see the piece of code to open the camera stream from a python notebook.

**Comment by m21c** [comment]
Hi Fabien, I'm coming across this same error but on another machine (macos Ventura, arm64). I followed the same instructions as in the Medium article you previously mentioned, but I'm not sure how to fix this again.. any suggestions? 

The error is now: 
`DllNotFoundException: libArm_kinematics.dylib assembly:<unknown assembly> type:<unknown type> member:(null)`
`JointServiceServer.Start () (at ./Library/PackageCache/com.pollenrobotics.reachysimulator@e146f91c5b/Scripts/JointServiceServer.cs:56)`

I ran the `otool` command to check the libraries, and this is the output:
`(base) ➜  Plugins otool -L libArm_kinematics.dylib`
`libArm_kinematics.dylib:`
`    @rpath/libArm_kinematics.dylib (compatibility version 0.0.0, current version 0.0.0)`
`    /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.23.0)`
`    /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)`

And also the `dyld_info` command:
`(base) ➜  Plugins dyld_info -dependents libArm_kinematics.dylib`
`libArm_kinematics.dylib [x86_64]:`
`    -dependents:`
`        attributes     load path`
`                       /usr/lib/libc++.1.dylib`
`                       /usr/lib/libSystem.B.dylib`

However, the libc++.1 lib is not in that path. I tried to install it with conda, and I found it in:
`/Users/username/opt/anaconda3/lib/libc++.1.0.dylib`

but when I copy it across (as described [here](https://github.com/pytorch/pytorch/issues/36941#issuecomment-617359369)), in unity I get the warnings:
`Cannot set deprecated build target 'OSXIntel' for Assets/ReachySimulator/Plugins/libc++.1.0.dylib`
`Cannot set deprecated build target 'OSXIntel64' for Assets/ReachySimulator/Plugins/libc++.1.0.dylib`

and the same DllNotFoundException error is still there when hitting the play button.

Could you please help me with this? 

Thank you so much!

**Comment by FabienDanieau** [comment]
Hi @m21c, libArm_kinematics.dylib is compiled for a x86_64 architecture that's why it is not working on your ARM64 system

**Comment by FabienDanieau** [comment]
Yes you can compile it for your system: https://github.com/pollen-robotics/Arm_kinematics
